aws-xray-auto-instrumentation-agent-runtime/src/integtest/java/com/amazonaws/xray/agent/handlers/downstream/MockHttpClient.java

```diff
@@ -0,0 +1,103 @@
+package com.amazonaws.xray.agent.handlers.downstream;
+
+import com.amazonaws.http.apache.client.impl.ConnectionManagerAwareHttpClient;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpVersion;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.conn.ClientConnectionManager;
+import org.apache.http.conn.HttpClientConnectionManager;
+import org.apache.http.entity.BasicHttpEntity;
+import org.apache.http.impl.io.EmptyInputStream;
+import org.apache.http.message.BasicHttpResponse;
+import org.apache.http.message.BasicStatusLine;
+import org.apache.http.params.HttpParams;
+import org.apache.http.protocol.HttpContext;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * TODO: When Disco removes its data accessor pattern in a future version, delete this & use a mock instead
+ */
+public class MockHttpClient implements ConnectionManagerAwareHttpClient {
+    private String responseContent;
+    private HttpUriRequest lastRequest; // Hacky way of spying on the last made request
```

Comment:
Just an observation, not sure if it applies here, but it tends to lead to simpler test code to use a mock server (e.g wiremock) instead of a mock client (was reminded of it by this comment ;) )

Reply:
In this case we're mocking the HTTP client used by the AWS SDK under the hood rather our own usage of an HTTP client, so I don't think it would apply in this case. But thanks for the pointer in general :)

Reply:
I think you're mocking the client to use a particular response, the `responseContent` and keep track of requests `lastRequest`. But you should be able to do the same thing by instead, stubbing a wiremock server and pointing the dynamodb client's endpoint to the localhost wiremock server. Least chance of disco-related drift of them all I think and no need for even `Whitebox` :)

I might still be missing something, let me know, but want to make sure on the same page.

Reply:
Wow, I actually think this would work. I just spent like a full day trying to force a solution around the old way we did this but I'm almost certain WireMock would be a much more elegant way of setting up these tests. Will include it in the next PR here - probably going to be a while while we figure out refactoring the repo to use Gradle and a Disco plugin-based architecture.

Reply:
@anuraaga I attempted to use WireMock by using this method before making any AWS SDK requests, instead of using the `MockHTTPClient` like I was before. I based it on your tests [here](https://github.com/aws/aws-xray-sdk-java/blob/master/aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/internal/UnsignedXrayClientTest.java):

```java
@Rule
public WireMockRule wireMockRule = new WireMockRule(wireMockConfig().dynamicPort());

// ...

    private void mockHttpClient(String responseContent) {
        stubFor(post(anyUrl())
                .willReturn(aResponse()
                        .withStatus(200)
                        .withBody(responseContent)));
    }
```

Though it failed with the following error, which I'm confused by since it seems like this is an error response from the DynamoDB service, but I don't know why we're even attempting to communicate with it while using WireMock. Going to stick with the mocking strategy we have for now, but this should still be possible with a little deeper dive.

```
    [junit] Testcase: testDynamoDBListTable(com.amazonaws.xray.agent.runtime.handlers.downstream.AWSHandlerIntegTest):	Caused an ERROR
    [junit] The security token included in the request is invalid. (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: UnrecognizedClientException; Request ID: VC4R7L2DNM56TRPJ24SB9GBKSFVV4KQNSO5AEMVJF66Q9ASUAAJG; Proxy: null)
    [junit] com.amazonaws.services.dynamodbv2.model.AmazonDynamoDBException: The security token included in the request is invalid. (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: UnrecognizedClientException; Request ID: VC4R7L2DNM56TRPJ24SB9GBKSFVV4KQNSO5AEMVJF66Q9ASUAAJG; Proxy: null)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1811)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1395)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1371)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1145)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:802)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:770)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:744)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:704)
    [junit] 	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:686)
    [junit] 	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:550)
    [junit] 	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:530)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.doInvoke$original$PCsIXk1Q(AmazonDynamoDBClient.java:5136)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.doInvoke$original$PCsIXk1Q$accessor$3Vc0ptxr(AmazonDynamoDBClient.java)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient$auxiliary$dv5mgWr6.call(Unknown Source)
    [junit] 	at software.amazon.disco.agent.awsv1.AWSClientInvokeInterceptor.doInvoke(AWSClientInvokeInterceptor.java:62)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.doInvoke(AmazonDynamoDBClient.java)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.invoke(AmazonDynamoDBClient.java:5103)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.executeListTables(AmazonDynamoDBClient.java:2484)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.listTables(AmazonDynamoDBClient.java:2450)
    [junit] 	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.listTables(AmazonDynamoDBClient.java:2496)
    [junit] 	at com.amazonaws.xray.agent.runtime.handlers.downstream.AWSHandlerIntegTest.testDynamoDBListTable(AWSHandlerIntegTest.java:124)
    [junit] 	at com.github.tomakehurst.wiremock.junit.WireMockRule$1.evaluate(WireMockRule.java:72)
```

Reply:
wiremock doesn't intercept requests magically, so you need to make sure you set the endpoint of the SDK client to localhost, which it looks like isn't happening. Here's an example

https://github.com/openzipkin/zipkin-aws/blob/master/brave-instrumentation/aws-java-sdk-v2-core/src/test/java/brave/instrumentation/awsv2/ITTracingExecutionInterceptor.java#L62